server/src/main/java/org/tctalent/server/service/db/impl/CandidateServiceImpl.java

```diff
@@ -405,2392 +405,2410 @@
+      Pattern p = Pattern.compile(linkedInRegex);
+      Matcher m = p.matcher(request.getLinkedInLink());
+      if (m.find()) {
+        candidate.setLinkedInLink(request.getLinkedInLink());
+      } else {
+        throw new InvalidRequestException("This is not a valid LinkedIn link.");
+      }
+    } else {
+      candidate.setLinkedInLink(null);
+    }
+    candidate.setAuditFields(candidate.getUser());
+    return save(candidate, true);
+  }
+
+  public void storeCandidateTaskAnswer(
+      QuestionTaskAssignment questionTaskAssignment, String answer)
+      throws InvalidRequestException {
+    Task task = questionTaskAssignment.getTask();
+    if (task instanceof QuestionTask) {
+      String answerField = ((QuestionTask) task).getCandidateAnswerField();
+      Candidate candidate = questionTaskAssignment.getCandidate();
+      if (answerField == null) {
+        //Store answer in a candidate property
+        String propertyName = task.getName();
+        candidatePropertyService.createOrUpdateProperty(
+            candidate, propertyName, answer, questionTaskAssignment);
+      } else {
+        //Store answer in the candidate field
+
+        //TODO JC This is a bit of a hack - we need a better way of processing things like candidateExams
+        String candidateExamsFieldName = "candidateExams";
+        if (answerField.startsWith(candidateExamsFieldName + ".")) {
+          //Special code for storing to candidate exams
+          //Extract the exam name from the answerField
+          String examName = answerField.substring(candidateExamsFieldName.length() + 1);
+          Exam examType = Exam.valueOf(examName);
+
+          //See if we already have an entry for this exam
+          Optional<CandidateExam> examO = Optional.empty();
+          final List<CandidateExam> candidateExams = candidate.getCandidateExams();
+          if (candidateExams != null) {
+            examO = candidateExams.stream().filter(
+                e -> e.getExam().equals(examType)).findFirst();
+          }
+
+          //Get or create the CandidateExam object
+          CandidateExam exam;
+          if (examO.isPresent()) {
+            exam = examO.get();
+          } else {
+            //Create new exam
+            exam = new CandidateExam();
+            exam.setCandidate(candidate);
+            exam.setExam(examType);
+            exam.setYear((long) OffsetDateTime.now().getYear());
+          }
+
+          //Update existing exam result
+          exam.setScore(answer);
+
+          //Save the exam
+          candidateExamRepository.save(exam);
         } else {
-            throw new InvalidRequestException("Task is not a QuestionTask: " + task.getName());
+          Object answerValue = convertAnswerToCorrectType(answerField, answer);
+          try {
+            PropertyUtils.setProperty(candidate, answerField, answerValue);
+          } catch (IllegalAccessException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 1249, 1252.